platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/ExceptionProblemRegistry.java

```diff
@@ -35,14 +40,119 @@
 @ServiceScope(Scope.BuildSession.class)
 public class ExceptionProblemRegistry {
 
-    private final Multimap<Throwable, Problem> problemsForThrowables = Multimaps.synchronizedMultimap(HashMultimap.<Throwable, Problem>create());
-
+    private final Multimap<Throwable, Problem> problemsForThrowables = Multimaps.synchronizedMultimap(MultimapBuilder.linkedHashKeys().linkedHashSetValues().<Throwable, Problem>build());
 
     public void onProblem(Throwable exception, Problem problem) {
         problemsForThrowables.put(exception, problem);
     }
 
-    public Map<Throwable, Collection<Problem>> getProblemsForThrowables() {
-        return problemsForThrowables.asMap();
+    public ProblemLookup getProblemLookup() {
+        return new DefaultProblemLookup();
+    }
+
+    /*
+     * Workaround for the fact that the exception thrown by the worker is not the same instance as the one that was thrown by the build. With the lookup we can find the original exception by comparing
+     * the stack frames. The comparison is expensive, so we only do when it's necessary (when the original exception does not contain the target and there's a matching class name and message).
+     */
+    private class DefaultProblemLookup implements ProblemLookup {
+
+
+        private final Multimap<String, Throwable> lookup;
+
+        DefaultProblemLookup() {
+            this.lookup = initLookup(problemsForThrowables.keySet());
+        }
+
+        private Multimap<String, Throwable> initLookup(Set<Throwable> exceptions) {
+            Multimap<String, Throwable> lookup = ArrayListMultimap.create();
+            for (Throwable exception : exceptions) {
+                lookup.put(key(exception), exception);
+            }
+            return lookup;
+        }
+
+        private String key(Throwable t) {
+            return t.getClass().getName() + ":" + messageOf(t);
+        }
+
+        private String messageOf(Throwable t) {
+            String result = "";
+            try {
+                String message = t.getMessage();
+                result = message == null ? "" : message;
+            } catch (RuntimeException ignore) {
+                // ignore exceptions with faulty getMessage() implementation
```

Comment:
❓ Won't this cause issues finding the proper exception if there are multiple instances of an exception with a faulty `getMessage()` implementation that are thrown?

Reply:
I don't see how. Both faulty exceptions would have still the same key in the lookup.